src/db/connection.cr

```diff
@@ -23,6 +23,47 @@ module DB
     include SessionMethods(Connection, Statement)
     include BeginTransaction
 
+    @ttl : Int32 = 600
+    @retry_delay : Int32 = 10
+
+    @pool : DB::Pool(self)?
+    @expire_at : Time?
+
+    getter expire_at
+
+    def expire!
+      return unless expired?
+      return unless pool = @pool
+      reaped = pool.replace_expired_resource!(self)
+
+      return if reaped || closed?
+
+      Reaper.reap!(@retry_delay) { expire! }
+    end
+
+    def expired?
+      expire_at.try &.<=(Time.utc)
+    end
+
+    def connect!(connected_pool : DB::Pool(self), margin : Int32 = 0)
+      return if @pool
+
+      timer = @ttl + (margin * 10)
+      @expire_at = Time.utc + timer.seconds
+      @pool = connected_pool
+
+      Reaper.reap!(timer) { expire! }
+    end
+
+    class Reaper
+      def self.reap!(sleep_for : Float64, &callback : -> _)
+        spawn(same_thread: true) do
+          sleep(sleep_for)
+          callback.call
```

Comment:
```suggestion
          loop do
            sleep(sleep_for)
            callback.call
          end
```
Might want to keep yourself in this loop so that any missed margins will cycle around past expiry, not sure though